hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusSubmitMessageHandler.java

```diff
@@ -280,6 +329,103 @@ public static byte[] noThrowSha384HashOf(final byte[] byteArray) {
         }
     }
 
+    /**
+     * Check if the submit message transaction is fee exempt
+     *
+     * @param feeExemptKeyList The list of keys that are exempt from fees
+     * @param keyVerifier The key verifier of this transaction
+     * @return if the transaction is fee exempt
+     */
+    private boolean isFeeExempted(@NonNull final List<Key> feeExemptKeyList, @NonNull final KeyVerifier keyVerifier) {
+        if (!feeExemptKeyList.isEmpty()) {
+            for (final var key : feeExemptKeyList) {
+                final var keyVerificationResult = keyVerifier.verificationFor(key);
+                if (keyVerificationResult.passed()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Validate that each topic custom fee has equal or lower value than the payer's limit
+     *
+     * @param topicCustomFees The topic's custom fee list
+     * @param payerCustomFeeLimits List with limits of fees that the payer is willing to pay
+     */
+    private void validateFeeLimits(
+            @NonNull final List<ConsensusCustomFee> topicCustomFees,
+            @NonNull final List<FixedFee> payerCustomFeeLimits,
+            @NonNull final HandleContext context) {
+        // Validate the duplication of payer custom fee limits
+        validateDuplicationFeeLimits(payerCustomFeeLimits);
+
+        // Extract the token fees and hbar fees from the topic custom fees
+        Map<TokenID, Long> tokenFees = new HashMap<>();
+        AtomicReference<Long> hbarFee = new AtomicReference<>(0L);
```

Comment:
We could re-use these maps to construct the synthetic crypto transfers right ?

Reply:
It is not possible to reuse these maps to build transfer bodies. There is no information about the collectors here, and the amount is the sum of all fees that have the same denomination.
Since we can have two or more custom fees with the same denomination, but with different collectors, we need to know the total amount per token and the total hbar amount. We use these maps to validate if the limits (`max_custom_fee`) that the payer is willing to pay are greater than or equal to the sum we calculate here.